src/calendar_duration.rs

```diff
@@ -0,0 +1,244 @@
+use core::num::NonZeroU32;
+
+use crate::{expect, try_opt};
+
+/// ISO 8601 duration type.
+///
+/// With this type a duration can be expressed as a combination of multiple components: months,
+/// days, minutes, seconds and nanoseconds.
+///
+/// # Examples
+///
+/// ```
+/// # use chrono::CalendarDuration;
+/// // Encoding 1½ year as a duration of a year and 6 months:
+/// let _duration = CalendarDuration::new().years_and_months(1, 6).unwrap();
+///
+/// // Encoding 1½ year as a duration of 12 months and 183 days (366 / 2):
+/// let _duration = CalendarDuration::new().months(12).days(183);
+///
+/// // Encoding 1½ year as a duration of 548 days (365 + 366 / 2):
+/// let _duration = CalendarDuration::new().days(548);
+///
+/// // Encoding 1½ year as a duration in seconds:
+/// let _duration = CalendarDuration::new().seconds(548 * 24 * 60 * 60);
+/// ```
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct CalendarDuration {
+    // Components with a nominal duration
+    months: u32,
+    days: u32,
+    // Components with an accurate duration
+    // `seconds` can either encode `minutes << 6 | seconds`, or just seconds.
+    seconds: u32,
+    // `nanos` encodes `nanoseconds << 2 | has_minutes << 1 | 1`.
+    nanos: NonZeroU32,
+}
+
+impl CalendarDuration {
+    /// Crate a new duration initialized to `0`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use chrono::CalendarDuration;
+    /// // Duration in calendar months and days.
+    /// CalendarDuration::new().months(15).days(5);
+    /// // Duration in calendar days, and 3 hours.
+    /// CalendarDuration::new().days(5).hms(3, 0, 0).unwrap();
+    /// // Duration that will precisely count the number of seconds, including leap seconds.
+    /// CalendarDuration::new().seconds(23_456).millis(789).unwrap();
+    /// ```
+    pub const fn new() -> Self {
+        Self {
+            months: 0,
+            days: 0,
+            seconds: 0,
+            nanos: expect!(NonZeroU32::new(1), "can't fail, non-zero"),
+        }
+    }
+
+    /// Set the months component of this duration to `months`.
+    pub const fn months(mut self, months: u32) -> Self {
```

Comment:
I think we should call this `with_months()` and allow `months()` to be used for the getter (same for the other ones).

Reply:
Renamed. I was going for having short names for initialization, but there is something to say for making the getters have the short names.